rpcperms/interceptor.go

```diff
@@ -165,6 +207,66 @@ func (r *InterceptorChain) Permissions() map[string][]bakery.Op {
 	return c
 }
 
+// RegisterMiddleware registers a new middleware that will handle request/
+// response interception for all RPC messages that are initiated with a custom
+// macaroon caveat. The name of the custom caveat a middleware is handling is
+// also its unique identifier. Only one middleware can be registered for each
+// custom caveat.
+func (r *InterceptorChain) RegisterMiddleware(mw *MiddlewareHandler) error {
+	r.Lock()
+	defer r.Unlock()
+
+	// For now we only want one middleware per custom caveat name. If we
+	// allowed multiple middlewares handling the same caveat there would be
+	// a need for extra call chaining logic and they could overwrite each
+	// other's responses.
+	registered, ok := r.registeredMiddleware[mw.customCaveatName]
+	if ok {
+		return fmt.Errorf("a middleware is already registered for the "+
+			"custom caveat name '%s': %v", mw.customCaveatName,
+			registered.middlewareName)
+	}
+
+	r.registeredMiddleware[mw.customCaveatName] = mw
+
+	return nil
+}
+
+// RemoveMiddleware removes the middleware that handles the given custom caveat
+// name.
+func (r *InterceptorChain) RemoveMiddleware(customCaveatName string) {
+	r.Lock()
+	defer r.Unlock()
+
+	delete(r.registeredMiddleware, customCaveatName)
+}
+
+// CustomCaveatSupported makes sure a middleware that handles the given custom
+// caveat name is registered. If none is, an error is returned, signalling to
+// the macaroon bakery and its validator to reject macaroons that have a custom
+// caveat with that name.
+//
+// NOTE: This method is part of the macaroons.CustomCaveatAcceptor interface.
+func (r *InterceptorChain) CustomCaveatSupported(customCaveatName string) error {
+	r.RLock()
+	defer r.RUnlock()
+
+	// We only accept requests with a custom caveat if we also have a
+	// middleware registered that handles that custom caveat. That is
+	// crucial for security! Otherwise a request with an encumbered (=has
+	// restricted permissions based upon the custom caveat condition)
+	// macaroon would not be validated against the limitations that the
+	// custom caveat implicate.
+	for _, middleware := range r.registeredMiddleware {
```

Comment:
Given it's a map, why not just look it up here directly?

Reply:
Good point, fixed.

Reply:
Still here in the latest commit?